### PR TITLE
net-misc/nextcloud-client: add 3.14.2, port to qt6

### PR DESCRIPTION
**Pull request by Flowdalic**
Closes: https://bugs.gentoo.org/926673

```$ diff -u nextcloud-client-3.1{3,4}.2.ebuild
--- nextcloud-client-3.13.2.ebuild	2024-08-22 15:09:29.950312105 +0200
+++ nextcloud-client-3.14.2.ebuild	2024-10-22 12:29:13.214885486 +0200
@@ -14,7 +14,7 @@
 
 LICENSE="CC-BY-3.0 GPL-2"
 SLOT="0"
-KEYWORDS="amd64 ~arm64 ~ppc64 ~x86"
+KEYWORDS="~amd64 ~arm64 ~ppc64 ~x86"
 IUSE="doc dolphin nautilus test webengine"
 RESTRICT="!test? ( test )"
 
@@ -23,31 +23,24 @@
 RDEPEND="
 	>=dev-db/sqlite-3.34:3
 	>=dev-libs/openssl-1.1.0:0=
-	dev-libs/qtkeychain:=[qt5(+)]
-	dev-qt/qtcore:5
-	dev-qt/qtdbus:5
-	dev-qt/qtdeclarative:5[widgets]
-	dev-qt/qtgui:5=
-	dev-qt/qtnetwork:5[ssl]
-	dev-qt/qtquickcontrols2:5
-	dev-qt/qtsql:5[sqlite]
-	dev-qt/qtsvg:5
-	dev-qt/qtwebsockets:5
-	dev-qt/qtwidgets:5
+	dev-libs/qtkeychain:=[qt6(+)]
+	dev-qt/qtbase:6[dbus,gui,network,sql,widgets]
+	dev-qt/qtdeclarative:6[widgets]
+	dev-qt/qtsvg:6
+	dev-qt/qtwebsockets:6
 	net-libs/libcloudproviders
-	kde-frameworks/karchive:5
+	kde-frameworks/karchive:6
 	sys-libs/zlib
 	dolphin? (
-		kde-frameworks/kcoreaddons:5
-		kde-frameworks/kio:5
+		kde-frameworks/kcoreaddons:6
+		kde-frameworks/kio:6
 	)
 	nautilus? ( dev-python/nautilus-python )
-	webengine? ( dev-qt/qtwebengine:5[widgets] )
+	webengine? ( dev-qt/qtwebengine:6[widgets] )
 "
 DEPEND="
 	${RDEPEND}
-	dev-qt/qtconcurrent:5
-	dev-qt/qtxml:5
+	dev-qt/qtbase:6[concurrent,xml]
 	|| ( gnome-base/librsvg media-gfx/inkscape )
 	doc? (
 		dev-python/sphinx
@@ -56,23 +49,27 @@
 		virtual/latex-base
 	)
 	test? (
-		dev-qt/qttest:5
 		dev-util/cmocka
 	)
 "
 BDEPEND="
-	dev-qt/linguist-tools:5
+	dev-qt/qttools:6[linguist]
 	dolphin? ( kde-frameworks/extra-cmake-modules )
 "
 
 PATCHES=(
 	"${FILESDIR}"/${PN}-3.6.6-no-redefine-fortify-source.patch
-	"${FILESDIR}"/${PN}-3.13.2-fix_build_kf6.patch
+	# https://github.com/nextcloud/desktop/pull/7383
+	"${FILESDIR}"/${PN}-3.14.2-doc-fix.patch
 )
 
 src_prepare() {
 	# Keep tests in ${T}
 	sed -i -e "s#\"/tmp#\"${T}#g" test/test*.cpp || die
+	# The image under images/building/path.png does no longer
+	# exist. This is already fixed upstream and can be dropped on the
+	# next bump
+	sed -i '/images\/building\/path.png/d' doc/building.rst || die
 
 	cmake_src_prepare
 }
@@ -83,8 +80,9 @@
 		-DBUILD_UPDATER=OFF
 		$(cmake_use_find_package doc Sphinx)
 		$(cmake_use_find_package doc PdfLatex)
-		$(cmake_use_find_package webengine Qt5WebEngine)
-		$(cmake_use_find_package webengine Qt5WebEngineWidgets)
+		$(cmake_use_find_package webengine Qt6WebEngine)
+		$(cmake_use_find_package webengine Qt6WebEngineWidgets)
+		-DBUILD_WITH_WEBENGINE=$(usex webengine)
 		-DBUILD_SHELL_INTEGRATION_DOLPHIN=$(usex dolphin)
 		-DBUILD_SHELL_INTEGRATION_NAUTILUS=$(usex nautilus)
 		-DBUILD_TESTING=$(usex test)
```

### CODE REVIEW

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	|| ( gnome-base/librsvg media-gfx/inkscape )
+	doc? (
+		dev-python/sphinx
+		dev-tex/latexmk
+		dev-texlive/texlive-latexextra
+		virtual/latex-base
+	)
+	test? (
+		dev-util/cmocka
+	)
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:56]: can be on one line then

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		kde-frameworks/kcoreaddons:6
+		kde-frameworks/kio:6
+	)
+	nautilus? ( dev-python/nautilus-python )
+	webengine? ( dev-qt/qtwebengine:6[widgets] )
+"
+DEPEND="
+	${RDEPEND}
+	dev-qt/qtbase:6[concurrent,xml]
+	|| ( gnome-base/librsvg media-gfx/inkscape )
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:47]: should be one separate line per dep

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+IUSE="doc dolphin nautilus test webengine"
+RESTRICT="!test? ( test )"
+
+# slot op for qtqui as this package uses private API parts of qtqui
+# src/gui/generalsettings.cpp:#include <private/qzipwriter_p.h>
+RDEPEND="
+	>=dev-db/sqlite-3.34:3
+	>=dev-libs/openssl-1.1.0:0=
+	dev-libs/glib:2
+	dev-libs/qtkeychain:=[qt6(+)]
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:27]: In theory you are leaving the possibility of someone having pre-Qt6 enabled dev-libs/qtkeychain installed to attempt to upgrade to nextcloud-client-3.14.2, which is bound to fail. These things happen, so you can't just depend on unversioned qtkeychain with `qt6(+)` assumption.
```suggestion
	>=dev-libs/qtkeychain-0.14.2:=[qt6(+)]
```

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+EAPI=8
+
+inherit cmake virtualx xdg
+
+DESCRIPTION="Desktop Syncing Client for Nextcloud"
+HOMEPAGE="https://github.com/nextcloud/desktop"
+SRC_URI="
+	https://github.com/nextcloud/desktop/archive/v${PV/_/-}.tar.gz -> ${P}.tar.gz
+"
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:12]: can be one line, this looks odd.

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	dev-qt/qtbase:6[dbus,gui,network,sql,widgets]
+	dev-qt/qtdeclarative:6[widgets]
+	dev-qt/qtsvg:6
+	dev-qt/qtwebsockets:6
+	net-libs/libcloudproviders
+	kde-frameworks/karchive:6
+	kde-frameworks/kguiaddons:6
+	sys-libs/zlib
+	dolphin? (
+		kde-frameworks/kcoreaddons:6
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:38]: Just to make sure, there was no minimum KF6 version required by build system?

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		dev-texlive/texlive-latexextra
+		virtual/latex-base
+	)
+	test? (
+		dev-util/cmocka
+	)
+"
+BDEPEND="
+	dev-qt/qttools:6[linguist]
+	dolphin? ( kde-frameworks/extra-cmake-modules )
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:60]: Similarly, you need a minimum ECM version that will guarantee Qt6/KF6 aware modules. Build system will likely define that, however, with current `ecm.eclass` baseline (5.106.0) you'll definitely be safe.
```suggestion
	dolphin? ( >=kde-frameworks/extra-cmake-modules-5.106.0 )
```

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+src_test() {
+	TEST_VERBOSE=1 virtx cmake_src_test
+}
+
+src_compile() {
+	local compile_targets=(all)
+	if use doc; then
+		compile_targets+=(doc doc-man)
+	fi
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:105]:
```suggestion
	local compile_targets=( all )
	if use doc; then
		compile_targets+=( doc doc-man )
	fi
```

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	# https://github.com/nextcloud/desktop/pull/7383
+	"${FILESDIR}"/${PN}-3.14.2-doc-fix.patch
+)
+
+src_prepare() {
+	# Keep tests in ${T}
+	sed -i -e "s#\"/tmp#\"${T}#g" test/test*.cpp || die
+	# The image under images/building/path.png does no longer
+	# exist. This is already fixed upstream and can be dropped on the
+	# next bump
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:74]:
```suggestion
	# The image under images/building/path.png does no longer exist.
	# This is already fixed upstream and can be dropped on the next bump.
```
**Can** it be dropped, though?

**Review by a17r — CHANGES_REQUESTED** [review]

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+	dev-libs/glib:2
+	dev-libs/qtkeychain:=[qt6(+)]
+	dev-qt/qt5compat:6
+	dev-qt/qtbase:6[dbus,gui,network,sql,widgets]
+	dev-qt/qtdeclarative:6[widgets]
+	dev-qt/qtsvg:6
+	dev-qt/qtwebsockets:6
+	net-libs/libcloudproviders
+	kde-frameworks/karchive:6
+	kde-frameworks/kguiaddons:6
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:35]: `k` before `n`
```suggestion
	kde-frameworks/karchive:6
	kde-frameworks/kguiaddons:6
	net-libs/libcloudproviders
```

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+KEYWORDS="~amd64 ~arm64 ~ppc64 ~x86"
+IUSE="doc dolphin nautilus test webengine"
+RESTRICT="!test? ( test )"
+
+# slot op for qtqui as this package uses private API parts of qtqui
+# src/gui/generalsettings.cpp:#include <private/qzipwriter_p.h>
+RDEPEND="
+	>=dev-db/sqlite-3.34:3
+	>=dev-libs/openssl-1.1.0:0=
+	dev-libs/glib:2
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:26]: `g` before `o`
```suggestion
	dev-libs/glib:2
	>=dev-libs/openssl-1.1.0:0=
```

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	dev-libs/qtkeychain:=[qt6(+)]
+	dev-qt/qt5compat:6
+	dev-qt/qtbase:6[dbus,gui,network,sql,widgets]
```
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:29]: Also, please check the comment on top of `RDEPEND` if the slot operator previously applied to `qtgui` is still necessary now on `qtbase`.

**Review by a17r — CHANGES_REQUESTED** [review]

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild (hunk 11 above)
Comment (a17r) [inline net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:29]: Where did `dev-qt/qtsql:5[sqlite]` go? is it no longer required?

**Review by a17r — CHANGES_REQUESTED** [review]

File: net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild (hunk 5 above)
Comment (Flowdalic) [reply net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:38]: It seems to require KF6 5.240. Not sure how to map this to Gentoo versions?

https://github.com/nextcloud/desktop/blob/v3.14.2/shell_integration/CMakeLists.txt#L15

https://github.com/nextcloud/desktop/blob/v3.14.2/shell_integration/dolphin/CMakeLists.txt#L8

Comment (a17r) [reply net-misc/nextcloud-client/nextcloud-client-3.14.2.ebuild:38]: That's KF6 development phase version, just use `6.0.0` then.

**Pull request closed without merging**